components/supervisor/frontend/src/ide/heart-beat.ts

```diff
@@ -70,6 +63,22 @@ export function schedule(wsInfo: WorkspaceInfo, sessionId: string): void {
         sendHeartBeat();
     }, activityInterval);
     toCancel.push(Disposable.create(() => clearInterval(intervalHandle)));
+
+    const handleUnload = async () => {
+        if (isSaaS) {
+            const result = await isSaaSServerGreaterThan("main.4091");
+            if (result) {
+                const unloadListener = () => { beaconWorkspacePageClose() }
+                window.addEventListener('unload', unloadListener);
+                toCancel?.push(Disposable.create(() => window.removeEventListener('unload', unloadListener)));
+                return;
+            }
+        }
+        const unloadListener = () => { sendHeartBeat(true) }
+        window.addEventListener('unload', unloadListener);
+        toCancel?.push(Disposable.create(() => window.removeEventListener('unload', unloadListener)));
+    }
+    handleUnload().catch(console.error);
```

Comment:
_Just a comment as I was scanning open PRs but did not mean to provide an in-depth review now._ 

I've no experience with `sendBeacon`, so had to read on it on https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon. 

This implementation has two problems:
1. usage of `unload` is discouraged, see docs
2. handler awaiting on fetch results before actually calling `sendBeacong` won't be effective, I guess. I didn't test it, but I'd bet the response would never be processed.

Reply:
@AlexTugarev  Thank you for your review 🧡 

> 1. usage of unload is discouraged, see docs

I tried with `pagehide` but it just not works, and `visibilitychange` seems not good for our use case, since switch tab should not send close event. 

> 2. handler awaiting on fetch results before actually calling `sendBeacon` won't be effective, I guess. I didn't test it, but I'd bet the response would never be processed.

the `async` keyword is not necessary, I'll remove it later